feat: differential scan for Pro engine support in pysemgrep

ihji opened the pull request:
This commit adds support for performing differential scans using the Pro engine within PySemgrep. Both the `scan` and `ci` commands have been updated to accommodate this functionality. PySemgrep will automatically switch to differential scan mode when both the `--baseline-commit` and `--pro` flags are specified in the command.

To further enhance flexibility, we have introduced an optional `--diff-depth` parameter. This parameter allows users to adjust the scanning speed, potentially sacrificing some accuracy in the process.

Test Plan:
  - Running diff scan validation script (modified binary path to pysemgrep and `--experimental` flag removed):
https://www.notion.so/semgrep/Differential-Scan-Support-for-Pro-Engine-4bdd04519e524aaeb616dc61b704e043?pvs=4#f02163357b5241c4a5704cee082c9d9b
  - Log:
```
Initialized empty Git repository in test_repos/go_diff_scan_test/tmp_test/.git/
[main (root-commit) 5e13806] commit deps
 3 files changed, 39 insertions(+)
 create mode 100644 test2.go
 create mode 100644 test3.go
 create mode 100644 test4.go
[main 861e69b] commit sink
 1 file changed, 13 insertions(+)
 create mode 100644 test.go
out of callee depth: OK
[main 20d4295] depth 2 callee
 1 file changed, 1 insertion(+), 1 deletion(-)
depth 2 callee: OK
HEAD is now at 861e69b commit sink
[main 6d48a25] depth 1 callee
 1 file changed, 1 insertion(+), 1 deletion(-)
depth 1 callee: OK
[main 2e5ad1c] remove pre-existing
 1 file changed, 9 insertions(+)
remove pre-existing: OK
[main 7ddb5ef] remove sources
 3 files changed, 3 insertions(+), 3 deletions(-)
no sources: OK
[main 4bb5d75] out of caller depth
 1 file changed, 1 insertion(+), 1 deletion(-)
out of caller depth: OK
[main 35a1425] depth 2 caller
 1 file changed, 1 insertion(+), 1 deletion(-)
depth 2 caller: OK
[main 58a0329] depth 1 caller
 1 file changed, 1 insertion(+), 1 deletion(-)
depth 1 caller: OK

```

  - actual command line per diff scan mode
```
> semgrep ci --baseline-commit HEAD~1 -d (per-file diff scan)

semgrep-core -json -rules /Users/heejong/.semgrep/semgrep_rules.json \
-targets /Users/heejong/.semgrep/semgrep_targets.txt -j 10 -timeout 30 -timeout_threshold 3 \
-max_memory 0 -json_time -fast

> semgrep ci --pro -d (inter-file whole scan)

semgrep-core-proprietary -json -rules /Users/heejong/.semgrep/semgrep_rules.json \
-targets /Users/heejong/.semgrep/semgrep_targets.txt -j 1 -timeout 30 -timeout_threshold 3 \
-max_memory 5000 -json_time -fast -deep_inter_file -timeout_for_interfile_analysis 10800 .

> semgrep ci --baseline-commit HEAD~1 --pro -d (intra-file diff scan)

semgrep-core-proprietary -json -rules /Users/heejong/.semgrep/semgrep_rules.json \
-targets /Users/heejong/.semgrep/semgrep_targets.txt -j 10 -timeout 30 -timeout_threshold 3 \
-max_memory 0 -json_time -fast -deep_intra_file

> semgrep ci --baseline-commit HEAD~1 --pro --diff-depth 2 -d (inter-file diff scan, depth 2)

semgrep-core-proprietary -json -rules /Users/heejong/.semgrep/semgrep_rules.json \
-diff_targets /Users/heejong/.semgrep/semgrep_diff_targets.txt -diff_depth 2 \
-targets /Users/heejong/.semgrep/semgrep_targets.txt -j 1 -timeout 30 -timeout_threshold 3 \
-max_memory 5000 -json_time -fast -deep_inter_file -timeout_for_interfile_analysis 10800 .

> semgrep ci --baseline-commit HEAD~1 --pro --diff-depth 0 -d (inter-file diff scan, depth 0)

semgrep-core-proprietary -json -rules /Users/heejong/.semgrep/semgrep_rules.json \
-diff_targets /Users/heejong/.semgrep/semgrep_diff_targets.txt -diff_depth 0 \
-targets /Users/heejong/.semgrep/semgrep_targets.txt -j 1 -timeout 30 -timeout_threshold 3 \
-max_memory 5000 -json_time -fast -deep_inter_file -timeout_for_interfile_analysis 10800 .
```

ihji committed:
feat: differential scan for Pro engine support in pysemgrep

nmote commented on cli/src/semgrep/commands/scan.py:
```diff
@@ -547,6 +548,16 @@ def convert(
+        help="""
+            The depth of the differential scan, the number of steps (both in the
+            caller and callee sides) from the targets in the call graph tracked
+            by the deep preprocessor. Only applied in differential scan mode.
```
I would make sure to mention pro and/or interfile here

ihji replied:
done

nmote commented:
If I'm not mistaken, the script you mention in your test plan calls osemgrep directly so it would not exercise this code at all. Could you clarify?

nmote commented on src/osemgrep/cli_scan/Scan_CLI.ml:
```diff
@@ -281,6 +282,18 @@ given baseline hash doesn't exist.
+  let info =
+    Arg.info [ "diff-depth" ]
+      ~doc:
+        {|The depth of the differential scan, the number of steps (both in the
+       caller and callee sides) from the targets in the call graph tracked
+       by the deep preprocessor. Only applied in differential scan mode.
+       Default to 2.
+       |}
```
Same feedback here as for the message in pysemgrep

ihji replied:
done

nmote reviewed (commented):
Could you add some more comments throughout and update the test plan? I'm having a hard time understanding what's going on here. I will take a closer look tomorrow.

ihji commented:
> If I'm not mistaken, the script you mention in your test plan calls osemgrep directly so it would not exercise this code at all. Could you clarify?

I used the same script but modified the path to pysemgrep and removed `--experimental` flag. So it called pysemgrep.

Good question. I've updated the PR description.

nmote commented on cli/src/semgrep/diff_scan_config.py:
```diff
@@ -0,0 +1,70 @@
+"""
+Differential scan configuration for Pro inter-file analysis
+
+It's important to note that the targeting schemes differ among three scanning
+modes: pro (interfile) differential scanning, open-source (per-file)
+differential scanning, and whole scan. Here's a breakdown of their respective
+targeting requirements:
+
+Pro Diff Scan: In this mode, all input files are designated as "targets", and
+specifically, the files that have changed between the head and baseline commits
+are considered "diff targets".
+
+OSS Diff Scan: For OSS differential scanning, the files that have changed
+between the head and baseline commits serve as the primary "targets".
+
+Whole Scan: In the case of the whole scan, all input files are categorized as
+"targets" without any distinction based on commit changes.
```
There are three kinds of scans here, but only two represented in the config below. It's not at all clear from reading the use sites that `is_diff_scan` implies Pro interfile. Could you clarify this, at least by renaming `DiffScan` etc. to `InterfileDiffScan`, or preferably by modeling the OSS diff scan with the same data structure?

ihji replied:
made an update to the PR. The `DiffScanConfig` has been renamed to `TargetModeConfig` and it now offers three options: "whole scan," "diff scan," and "pro diff scan."

nmote reviewed (commented):
Okay, I have a better handle on this but would still like to take another look once you've made some changes. I have one comment inline which I think will make this easier to understand.

I am also worried that the implication here is that when this ships, everyone running `semgrep ci` with the Pro flag turned on in the app will suddenly start getting the differential interfile analysis instead of intraproc. I'd be a lot more comfortable if it was opt-in to begin with, so we can have a few customers try it out and give feedback before rolling it out to everyone. Could you add a temporary flag and gate this behavior behind it?

Also, could you test that `semgrep ci` does in fact behave as I describe above? You should be able to set up a CI job using the docker image generated for this PR.

nmote reviewed (approved):
Nice, this is a lot clearer! Could you test that `semgrep ci` by default does an intrafile analysis now, and update the test plan accordingly?

ihji commented:
> Nice, this is a lot clearer! Could you test that `semgrep ci` by default does an intrafile analysis now, and update the test plan accordingly?

updated!
